dvc/repo/diff.py

```diff
@@ -85,3 +47,63 @@ def _exists(output):
     }
 
     return ret if any(ret.values()) else {}
+
+
+def _paths_checksums(repo):
+    """
+    A dictionary of checksums addressed by relpaths collected from
+    the current tree outputs.
+
+    To help distinguish between a directory and a file output,
+    the former one will come with a trailing slash in the path:
+
+        directory: "data/"
+        file:      "data"
+    """
+
+    return dict(_output_paths(repo))
+
+
+def _output_paths(repo):
+    on_working_tree = isinstance(repo.tree, LocalTree)
+
+    def _exists(output):
+        if on_working_tree:
+            return output.exists
+        return True
+
+    def _to_path(output):
+        return (
+            str(output)
+            if not output.is_dir_checksum
+            else os.path.join(str(output), "")
+        )
+
+    def _to_checksum(output):
+        if on_working_tree:
+            return repo.cache.local.tree.get_hash(output.path_info).value
+        return output.hash_info.value
+
+    for stage in repo.stages:
+        for output in stage.outs:
+            if _exists(output):
+                yield _to_path(output), _to_checksum(output)
+                if output.is_dir_checksum:
+                    yield from _dir_output_paths(repo, output, on_working_tree)
+
+
+def _dir_output_paths(repo, output, on_working_tree):
+    from dvc.config import NoRemoteError
+
+    try:
+        repo_tree = RepoTree(repo, stream=True)
+        for fname in repo_tree.walk_files(output.path_info):
+            if on_working_tree:
+                hash_ = repo.cache.local.tree.get_hash(fname).value
```

Comment:
We cannot use `repo_tree` because of the `state`, right?

Reply:
`repo.diff` is wrapped with `@lock` so we shouldn't have issues using `state` here I think?

Reply:
I guess, I was trying to point out that `BaseTree` uses `StateNoop`. So, subsequent operations might become slower? I am not quite sure though.
Anyway, the question was if we could replace this `repo.cache.local.tree` with just `repo_tree.get_hash`?